libs/core/langchain_core/tracers/core.py

```diff
@@ -533,39 +537,52 @@ def _on_llm_new_token(
         chunk: Optional[Union[GenerationChunk, ChatGenerationChunk]],
     ) -> Union[None, Coroutine[Any, Any, None]]:
         """Process new LLM token."""
+        return None
 
     def _on_llm_end(self, run: Run) -> Union[None, Coroutine[Any, Any, None]]:
         """Process the LLM Run."""
+        return None
 
     def _on_llm_error(self, run: Run) -> Union[None, Coroutine[Any, Any, None]]:
         """Process the LLM Run upon error."""
+        return None
 
     def _on_chain_start(self, run: Run) -> Union[None, Coroutine[Any, Any, None]]:
         """Process the Chain Run upon start."""
+        return None
 
     def _on_chain_end(self, run: Run) -> Union[None, Coroutine[Any, Any, None]]:
         """Process the Chain Run."""
+        return None
 
     def _on_chain_error(self, run: Run) -> Union[None, Coroutine[Any, Any, None]]:
         """Process the Chain Run upon error."""
+        return None
 
     def _on_tool_start(self, run: Run) -> Union[None, Coroutine[Any, Any, None]]:
         """Process the Tool Run upon start."""
+        return None
 
     def _on_tool_end(self, run: Run) -> Union[None, Coroutine[Any, Any, None]]:
         """Process the Tool Run."""
+        return None
 
     def _on_tool_error(self, run: Run) -> Union[None, Coroutine[Any, Any, None]]:
         """Process the Tool Run upon error."""
+        return None
 
     def _on_chat_model_start(self, run: Run) -> Union[None, Coroutine[Any, Any, None]]:
         """Process the Chat Model Run upon start."""
+        return None
 
     def _on_retriever_start(self, run: Run) -> Union[None, Coroutine[Any, Any, None]]:
         """Process the Retriever Run upon start."""
+        return None
 
     def _on_retriever_end(self, run: Run) -> Union[None, Coroutine[Any, Any, None]]:
         """Process the Retriever Run."""
+        return None
 
     def _on_retriever_error(self, run: Run) -> Union[None, Coroutine[Any, Any, None]]:
         """Process the Retriever Run upon error."""
+        return None
```

Comment:
this doesn't make a ton of sense, its either an abstract method or it isn't

Reply:
seems like separate issue from this pr (having nothing is equivalent to returning None)

Reply:
See previous comment: https://github.com/langchain-ai/langchain/pull/25520/files/dd1df352ffb2731807f10224b0919039782939d7#r1720771953
It's not an abstract method, but ruff thinks it should be if we don't add `return None`.
I can add `noqa` instead if you prefer.